modules/h2/src/test/java/org/h2/test/synth/TestChooseIndexWithOrderBy.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2004-2019 H2 Group. Multiple-Licensed under the MPL 2.0,
+ * and the EPL 1.0 (http://h2database.com/html/license.html).
+ * Initial Developer: H2 Group
+ */
+package org.h2.test.synth;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.h2.util.StringUtils;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import static org.junit.Assert.assertTrue;
+
+/**
+ * A test that runs random join statements against two databases and compares the results.
+ */
+public class TestChooseIndexWithOrderBy {
+    private static final int ROWS_CNT = 100;
+
+    private static Connection connection;
+
+    /**
+     * @throws SQLException On error.
+     */
+    @BeforeClass
+    public static void init() throws SQLException {
+        connection = DriverManager.getConnection("jdbc:h2:mem:idx_order_by");
+        sql("DROP TABLE IF EXISTS TEST");
+
+        sql("CREATE TABLE TEST (ID INT PRIMARY KEY, VAL0 INT, VAL1 INT, VAL2 VARCHAR)");
+        sql("CREATE INDEX IDX_VAL0 ON TEST(VAL0, ID)");
+        sql("CREATE INDEX IDX_VAL0_VAL1 ON TEST(VAL0, VAL1, ID)");
+
+        for (int i = 0; i < ROWS_CNT; ++i) {
+            sql(
+                "INSERT INTO TEST (ID, VAL0, VAL1, VAL2) VALUES (?, ?, ?, ?)",
+                i,
+                i / 10,
+                i,
+                "val" + i);
+        }
+    }
+
+    /**
+     * @throws SQLException On error.
+     */
+    @AfterClass
+    public static void cleanup() throws SQLException {
+        sql("DROP TABLE IF EXISTS TEST");
+
+        connection.close();
+    }
+
+    /**
+     * Check query plan. HASH_JOIN_IDX index must be chosen.
+     * @throws Exception On error.
+     */
+    @Test
+    public void test() throws Exception {
+        String plan = null;
+
+        plan = sqlStr("EXPLAIN SELECT VAL0, VAL1, VAL2 FROM TEST " +
+            "WHERE VAL0 = 0 " +
+            "ORDER BY VAL0, VAL1");
+        assertTrue("Unexpected plan: " + plan, plan.contains("IDX_VAL0_VAL1"));
+
+        plan = sqlStr("EXPLAIN SELECT VAL0, VAL1, VAL2 FROM TEST " +
+            "WHERE VAL0 = 0 " +
+            "ORDER BY 1, 2");
+        assertTrue("Unexpected plan: " + plan, plan.contains("IDX_VAL0_VAL1"));
+    }
+
+    /**
+     * @param sql SQL query.
+     * @param params Parameters.
+     * @throws SQLException On error.
+     * @return Result set or updated count are printed to string.
+     */
+    private static String sqlStr(String sql, Object... params) throws SQLException {
```

Comment:
Can we move sqlStr to some utils class? We have exactly the same code in TestHashJoin class and pretty the same in many others sql tests.